flang/docs/tutorials/addingIntrinsics.md

```diff
@@ -0,0 +1,383 @@
+# Adding and Implementing an Intrinsic Procedure in f18
+This document first describes where and how to implement an intrinsic and to write tests
+at the four different stages listed below. It then illustrates this with the example of
+Trim intrinsic. Readers that know support in the front-end part is done should only
+focus on the runtime and lowering related sections.
+
+The first step should be to be sure to have a copy of the standard and have
+read the requirements for the intrinsic to be implemented.
+In Fortran 2018, these are listed in section 16.9.
+Writing some end-to-end Fortran test cases and running them with existing compilers is
+a good way to get familiar with it, and will be useful to later write regression tests.
+
+To add a completely new intrinsic to f18, up to 4 steps might be needed:
+
+- Adding support in name resolution and semantic check
+- Adding support in front-end folding (only if intrinsic can be in constant expressions)
+- Adding support in the runtime (if runtime is the chosen way to implement the intrinsic)
+- Adding support in lowering
+
+For most of the Fortran 2018 intrinsics, the support in the first two steps (front-end),
+have been completed. It is very likely only lowering and runtime support is missing.
+
+
+## Quick checking the status of an intrinsic.
+This font-end support can be easily tested by running `f18 -funparse` on a test using the
+intrinsic and using an explicit INTRINSIC attribute statement to declare it. If an
+error about unknown intrinsic is raised, the front-end support is missing. Otherwise,
+name resolution is done.
+
+To test front-end folding, if applicable, the intrinsic can be used in the initialization
+of a PARAMETER. With the `f18 -funparse`, it should be observed that the parameter
+initialization has been folded. If an error is thrown, the front-end  folding support
+is missing.
+
+To test lowering, `bbc -emit-fir` should raise no error about missing intrinsics on the test
+case. Note that if the test case uses modern Fortran features, it is likely that it would
+currently hit other TODOs. So it is better to keep the test cases simple at first.
+
+To test the runtime support, it would be needed to try to compile and link the test program with the runtime.
```

Comment:
s/would/will/ (I'd try to keep consistent present tense throughout, since we hope they'll read this before or while doing it.)